sklearn/metrics/_classification.py

```diff
@@ -959,6 +959,24 @@ def matthews_corrcoef(y_true, y_pred, *, sample_weight=None):
         prediction, 0 an average random prediction and -1 and inverse
         prediction).
 
+    Notes
+    -----
+    :func:`matthews_corrcoef` is ill-defined (due to zero division)
+    when only one class is present in either the true or predicted
+    labels.
+
+    If only one of the true or predicted labels contains a single
+    class, the limit value is 0. This is sensible as it suggests
+    that the model either provided constant predictions on
+    non-constant data, or variable predictions on single-class data.
+    In such cases, the metric will return a value of 0.
+
+    However, if both the true and predicted labels contain only a
+    single class, the limit does not exist, rendering the metric
+    undefined. Consequently, in this scenario, the metric will
+    return a nan value. This behavior was chosen to avoid
+    returning 0 for perfect predictions on single-class data.
```

Comment:
> and returning a 1 for perfect predictions on single-class data  

But the definition above prays
```
The Matthews correlation coefficient (+1 represents a perfect
        prediction, 0 an average random prediction and -1 and inverse
        prediction).
```

I think we should add a `zero_division` param as proposed in the Issue, so the user decides what's returned for this case. For the case where only one of them has single class, I'm with you on what's sensible. But I could see people arguing otherwise 🤷‍♂️

Reply:
I agree that the desired behavior for perfect correlation on single class data is arguable. 

Some will say it's perfect prediction and should return 1, some will say it's zero division so should return 0 and raise a warning with an option to choose other return values via `zero_division`, to be in line with the other metrics.

My logic was:
- The two options of returning 0 or 1 are both mathematically (actual value and limit value) not correct.
- Differentiation between the non-existing limit case (perfect correlation for trivial task - we can't tell if the model is good) and the 0 limit case (non uniform predictions for single class data or the opposite - the model is not good) seems important.
-  `scikit-learn` users come from different backgrounds, while returning an arbitrary value and raising a warning might be better for SW engineers, returning nan for an undetermined value seems more like a lingua franca between developers, scientists, statisticians, teachers, students, etc.

The implementation is simple in either case, it's more about style and about how opinionated the module should be.

Reply:
precisely because not everyone will want the same, I'd argue a `zero_division` param gives flexibility. But let's find some other opinion, esp. from a code owner

Reply:
I also agree that we should have the `zero_division` parameter as we previously decided. @Redjest You can actually refer to the work of @marctorsoc in #23183 where the changes were already done.

Reply:
Alright, so the conclusion is that we want the zero limit cases (correspond to bad predictions) and the non-existing limit case (perfect prediction on single class data) to have the same behavior, determined by `zero_division` as in the other metrics?

Reply:
Yes indeed.